### ISSUE TITLE
MSVC compiler error with /std:c++latest

### ISSUE THREAD

**Issue by ysangkok** [boostorg/numeric_conversion#5]
Using the newest Visual Studio 2017 and its C++ compiler:

The errors mean 
* ',' is missing before ''
*  *Siehe Verweis auf die Klasse Vorlage-  Instanziierung*:  See reference to the Class Template-Instantiation
* *Schlüsselwort "typename" ist in* : Keyword 'typename' is in

Could this be a compiler error?

```
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(453): error C2143: Syntaxfehler: Es fehlt "," vor "
<" [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
  C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(467): note: Siehe Verweis auf die Klasse Vorlage-
  Instanziierung "boost::numeric::convdetail::trivial_converter_impl<Traits>", die kompiliert wird.
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(453): error C2518: Schlüsselwort "typename" ist in
der Basisklassenliste ungültig und wird ignoriert [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(454): error C2518: Schlüsselwort "typename" ist in
der Basisklassenliste ungültig und wird ignoriert [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(474): error C2143: Syntaxfehler: Es fehlt "," vor "
<" [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
  C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(497): note: Siehe Verweis auf die Klasse Vorlage-
  Instanziierung "boost::numeric::convdetail::rounding_converter<Traits,RangeChecker,RawConverter,Float2IntRounder>", d
  ie kompiliert wird.
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(474): error C2518: Schlüsselwort "typename" ist in
der Basisklassenliste ungültig und wird ignoriert [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(475): error C2518: Schlüsselwort "typename" ist in
der Basisklassenliste ungültig und wird ignoriert [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(504): error C2143: Syntaxfehler: Es fehlt "," vor "
<" [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
  C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(526): note: Siehe Verweis auf die Klasse Vorlage-
  Instanziierung "boost::numeric::convdetail::non_rounding_converter<Traits,RangeChecker,RawConverter>", die kompiliert
   wird.
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(504): error C2518: Schlüsselwort "typename" ist in
der Basisklassenliste ungültig und wird ignoriert [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
C:\local\boost_1_64_0\boost/numeric/conversion/detail/converter.hpp(505): error C2518: Schlüsselwort "typename" ist in
der Basisklassenliste ungültig und wird ignoriert [C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
C:\local\boost_1_64_0\boost/algorithm/string/detail/case_conv.hpp(33): error C2143: Syntaxfehler: Es fehlt "," vor "<"
[C:\Users\jtroelsen\ifcconverter\ifcconverter.vcxproj]
```

https://github.com/boostorg/numeric_conversion/blob/boost-1.64.0/include/boost/numeric/conversion/detail/converter.hpp#L453

**Comment by PatrickMGrace** [comment]
I am getting what seems to be the same error:
...\boost\numeric\conversion\detail\converter.hpp(453): error C2143: syntax error: missing ',' before '<'

**Comment by BurningEnlightenment** [comment]
No, this isn't a compiler error. If you look at the code you will notice, that the errors stem from the usage of `std::unary_function` which was removed in c++17 and MSVC doesn't provide the class anymore when compiling with `/std:c++latest`.
This was fixed with 593aafab5a7aba86f14df5f9a14a53fc4353bb00 and for an unknown reason reverted with 2009f0d4fc49a6db59b950a3140f8d1f58e1a8c6.
@brandon-kohn can you provide some insight, why it was reverted and when we can expect c++17 compliance?

**Comment by brandon-kohn** [comment]
Fix is up in master now. Please re-open if there is some lingering issue.

**Comment by DoumanAsh** [comment]
Is it included into current release (1.65.1)?

**Comment by PBonnema** [comment]
@DoumanAsh I have this problem to but I don't see the actual header files on the repo. Am I missing something? Same for the development branch.

**Comment by DoumanAsh** [comment]
What do you mean not seeing actual header files?
They lies by `include/boost`

**Comment by PBonnema** [comment]
I see I'm mistaken. I'm new to boost and I assumed that the boost project contained all libraries but it doesn't. You have to clone all libs you want separately. Mystery solved.

**Comment by rathaROG** [comment]
I'm new to boost and C++ and I have this problem too. So what should I do? Download and replace? Or I have to build something?

**Comment by BurningEnlightenment** [comment]
FYI @rathaROG the possible workarounds are summed up [here](https://www.reddit.com/r/cpp/comments/67vesx/has_the_boost_library_issue_with_visual_studio/dgtjytj/)

**Comment by rathaROG** [comment]
Thank you so much @BurningEnlightenment !

**Comment by Elypson** [comment]
In Boost 1_67_0, the error exists (again!?)

**Comment by brandon-kohn** [comment]
I don't have any issues compiling on msvc-14.1. Could one of you post a small example reproducing your issue?

**Comment by travnick** [comment]
I guess by `/std:c++latest` they mean `/std:c++17` which is available since `MSVC++ 14.11` (`Visual Studio 2017 version 15.3`)
